server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java

```diff
@@ -381,11 +407,12 @@ final ObjectMapper getDynamicObjectMapper(String name) {
      * because for dynamic mappings, a new field can be either mapped
      * as runtime or indexed, but never both.
      */
-    final void addDynamicRuntimeField(RuntimeField runtimeField) {
-        if (newFieldsSeen.add(runtimeField.name())) {
-            mappingLookup.checkFieldLimit(indexSettings().getMappingTotalFieldsLimit(), newFieldsSeen.size());
+    final boolean addDynamicRuntimeField(RuntimeField runtimeField) {
+        if (dynamicRuntimeFields.containsKey(runtimeField.name()) == false) {
+            mappingLookup.checkFieldLimit(indexSettings().getMappingTotalFieldsLimit(), dynamicRuntimeFields.size() + 1);
         }
-        dynamicRuntimeFields.add(runtimeField);
+        dynamicRuntimeFields.computeIfAbsent(runtimeField.name(), k -> new ArrayList<>(1)).add(runtimeField);
```

Comment:
Is this addressing counting for dynamic runtime fields? Can you expand on why this change was necessary?

Reply:
I've fixed this and added a test case.

Reply:
++ thanks!